File: Tff.FSharp.CodeStock/Tff.Ransom.VB/DollarProvider.vb
Public Class DollarProvider

    Private Shared Function GetSerialNumber() As [String]
        Dim random As New Random()
        Dim serialNumber As [String] = [String].Empty
        For j As Integer = 0 To 8
            serialNumber += random.[Next](0, 9).ToString()
        Next
        Return serialNumber
    End Function

    Public Function GetDollars(numberOfDollars As Int32) As List(Of Dollar)
        Dim dollars As New List(Of Dollar)()
        Dim currentDollar As Dollar = Nothing
        Dim random As New Random()
        For i As Integer = 0 To numberOfDollars - 1
            currentDollar = New Dollar()
            currentDollar.Id = i
            currentDollar.FederalReserveDistrict = random.[Next](1, 13)
            currentDollar.SeriesDate = 2000 + random.[Next](0, 10)
            dollars.Add(currentDollar)
        Next
        Dim serialNumbers As New Dictionary(Of Int32, [String])()
        For i As Integer = 0 To numberOfDollars - 1
            serialNumbers.Add(i, GetSerialNumber())
        Next
        For Each dollar As Dollar In dollars
            Dim currentId As Int32 = dollar.Id
            For Each serialNumber As KeyValuePair(Of Int32, [String]) In serialNumbers
                If currentId = serialNumber.Key Then
                    dollar.SerialNumber = serialNumber.Value
                    Exit For
                End If
            Next
        Next
        Return dollars
    End Function
End Class

File: Tff.FSharp.CodeStock/Tff.Ransom.FS/DollarProvider.fs
namespace Tff.Ransom.FS

open System
 
type dollar =
    {id: int;
     serialNumber: string;
     federalReserveDistrict: int;
     seriesDate: int;
     signature: string}

type dollarProvider() =
    let randomNumberGenerator = new System.Random()
    
    let createSerialNumber id = 
        List.init 9 (fun _ -> randomNumberGenerator.Next(0,9))
                                |> Seq.map string
                                |> String.concat ""
    let createDollar id serialNumber=
        let returnDollar = {id=id;
            serialNumber= serialNumber;
            federalReserveDistrict=randomNumberGenerator.Next(0,13);
            seriesDate=2000+randomNumberGenerator.Next(0,9);
            signature=""}
        returnDollar

    member this.GetDollars (numberOfDollars:int) =
        let dollars = List.init numberOfDollars (fun index -> createDollar index)
        let serialNumbers = List.init numberOfDollars (fun index -> createSerialNumber index)
        List.zip dollars serialNumbers
